File: Creational_Patterns/C++/FactoryMethod.cpp
#include <iostream>
#include <string>
#include <memory>

// abstract class
class Car
{
public:
	virtual ~Car() {};
	virtual void engineStart() = 0;
	virtual void engineStop() = 0;
};

class Porsche : public Car
{
public:
	virtual ~Porsche() {}

	void engineStart() override
	{
		std::cout << "Porsche started!" << std::endl;
	}

	void engineStop() override
	{
		std::cout << "Porsche stopped!" << std::endl;
	}

};

class ModelS : public Car
{
public:
	virtual ~ModelS() {};

	void engineStart() override
	{
		std::cout << "ModelS started!" << std::endl;
	}

	void engineStop() override
	{
		std::cout << "ModelS stopped!" << std::endl;
	}
};

enum Fuel {
	Electrical,
	Gasoline
};

class CarFactory
{
public:
	static std::unique_ptr<Car> getInstance(Fuel fuel)
	{
		if (fuel == Fuel::Gasoline)
		{
			return std::make_unique<Porsche>();
		}
		else if (fuel == Fuel::Electrical)
		{
			return std::make_unique<ModelS>();
		}
		else
		{
			return nullptr;
		}
	}
};


int main()
{
	std::cout << ":: Factory Method ::\n" << std::endl;

	std::unique_ptr<Car>  gasolineCar = CarFactory::getInstance(Fuel::Gasoline);
	gasolineCar->engineStart();

	std::unique_ptr<Car> electricalCar = CarFactory::getInstance(Fuel::Electrical);
	electricalCar->engineStart();

	gasolineCar->engineStop();
	electricalCar->engineStop();

	system("PAUSE");

	return 0;
}

File: Creational_Patterns/Java/FactoryMethod.java
abstract class Car {

	abstract void engineStart();

	abstract void engineStop();

}

class Porsche extends Car {

	public void engineStart() {
		System.out.println("Porsche started!");
	}

	public void engineStop() {
		System.out.println("Porsche stopped!");
	}

}

class ModelS extends Car {

	public void engineStart() {
		System.out.println("ModelS started!");
	}

	public void engineStop() {
		System.out.println("ModelS stopped!");
	}

}

enum Fuel {
	Electrical,
	Gasoline
};

class CarFactory {

	public static Car getInstance(Fuel fuel) {
		if (fuel == Fuel.Gasoline) {
			return new Porsche();
		} else if (fuel == Fuel.Electrical) {
			return new ModelS();
		}

		return null;
	}

}

class Main {

	public static void main(String[] args) {
		System.out.println(":: Factory Method ::\n");

		Car gasolineCar = CarFactory.getInstance(Fuel.Gasoline);
		gasolineCar.engineStart();

		Car electricalCar = CarFactory.getInstance(Fuel.Electrical);
		electricalCar.engineStart();

		gasolineCar.engineStop();
		electricalCar.engineStop();
	}

}
